test/Identity.Test/IdentityServer/BaseRequestValidatorTests.cs

```diff
@@ -0,0 +1,401 @@
+                                          |-> SetErrorResult
+    */
+    [Theory, BitAutoData]
+    public async Task ValidateAsync_IsBot_UserNotNull_ShouldBuildErrorResult_ShouldLogFailedLoginEvent(
+        [AuthFixtures.ValidatedTokenRequest] ValidatedTokenRequest tokenRequest,
+        CustomValidatorRequestContext requestContext,
+        GrantValidationResult grantResult)
+    {
+        // Arrange
+        var context = CreateContext(tokenRequest, requestContext, grantResult);
+
+        context.CustomValidatorRequestContext.CaptchaResponse.IsBot = true;
+        _sut.isValid = true;
+
+        // Act
+        await _sut.ValidateAsync(context);
+
+        var errorResponse = (ErrorResponseModel)context.GrantResult.CustomResponse["ErrorModel"];
+
+        // Assert        
+        await _eventService.Received(1)
+                           .LogUserEventAsync(context.CustomValidatorRequestContext.User.Id,
+                                             Core.Enums.EventType.User_FailedLogIn);
+        Assert.True(context.GrantResult.IsError);
+        Assert.Equal("Username or password is incorrect. Try again.", errorResponse.Message);
+    }
+
+    /* Logic path
+    ValidateAsync -> UpdateFailedAuthDetailsAsync -> _mailService.SendFailedLoginAttemptsEmailAsync
+                 |-> BuildErrorResultAsync -> _eventService.LogUserEventAsync
+                            (self hosted) |-> _logger.LogWarning() 
+                                          |-> SetErrorResult
+    */
+    [Theory, BitAutoData]
+    public async Task ValidateAsync_ContextNotValid_SelfHosted_ShouldBuildErrorResult_ShouldLogWarning(
+        [AuthFixtures.ValidatedTokenRequest] ValidatedTokenRequest tokenRequest,
+        CustomValidatorRequestContext requestContext,
+        GrantValidationResult grantResult)
+    {
+        // Arrange
+        var context = CreateContext(tokenRequest, requestContext, grantResult);
+        context.CustomValidatorRequestContext.CaptchaResponse.IsBot = false;
+        _globalSettings.Captcha.Returns(new GlobalSettings.CaptchaSettings());
+        _globalSettings.SelfHosted = true;
+        _sut.isValid = false;
+
+        // Act
+        await _sut.ValidateAsync(context);
+
+        // Assert
+        //? This is smelly, but the test suite wouldn't pass the IpAddress in properly so we just check for this message
+        _logger.Received(1).LogWarning(Constants.BypassFiltersEventId, "Failed login attempt. ");
+        var errorResponse = (ErrorResponseModel)context.GrantResult.CustomResponse["ErrorModel"];
+        Assert.Equal("Username or password is incorrect. Try again.", errorResponse.Message);
+    }
+
+    /* Logic path
+    ValidateAsync -> UpdateFailedAuthDetailsAsync -> _mailService.SendFailedLoginAttemptsEmailAsync
+                 |-> BuildErrorResultAsync -> _eventService.LogUserEventAsync 
+                                          |-> SetErrorResult
+    */
```

Comment:
This is awesome. Thank you.

Reply:
I was getting lost in the forest needed some bread crumbs

Reply:
I felt that trying to track what was being tested.